### PR TITLE
fix(dot/network): close notifications streams

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Changes

- close notifications streams for reading/writing when outbound/inbound
respectively

## Tests

```

```

## Issues

- https://github.com/ChainSafe/gossamer/issues/2046

## Primary Reviewer

- @noot

### CODE REVIEW

File: dot/network/notifications.go [hunk 1]
```diff
@@ -152,9 +152,12 @@ func createDecoder(info *notificationsProtocol, handshakeDecoder HandshakeDecode
...
-func (s *Service) createNotificationsMessageHandler(info *notificationsProtocol,
+func (s *Service) createNotificationsMessageHandler(
+	info *notificationsProtocol,
 	messageHandler NotificationsMessageHandler,
-	batchHandler NotificationsMessageBatchHandler) messageHandler {
+	batchHandler NotificationsMessageBatchHandler,
+) messageHandler {
+
 	return func(stream libp2pnetwork.Stream, m Message) error {
 		if m == nil || info == nil || info.handshakeValidator == nil || messageHandler == nil {
```
Comment (EclesioMeloJunior) [inline dot/network/notifications.go:162]: is it possible to add a defer to close this stream?

```go
return func(stream libp2pnetwork.Stream, m Message) error {
  defer func(){
    if err := stream.CloseWrite(); err != nil {
      logger.Tracef("failed to close stream for writing: %s", err)
    }
  }()
}
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/network/notifications.go [hunk 2]
```diff
@@ -201,6 +204,7 @@ func (s *Service) createNotificationsMessageHandler(info *notificationsProtocol,
 				info.inboundHandshakeData.Store(peer, hsData)
 
 				// once validated, send back a handshake
+				// TODO: getHandshake creates a new handshake struct. Rename?
```
Comment (noot) [inline dot/network/notifications.go:207]: `getHandshake` is just some function that satisfies the handshake getter function signature, it doesn't necessarily make a new struct, and could be different for each notifications protocol

**Review by noot — APPROVED** [review]
looks good!

**Review by edwardmack — APPROVED** [review]
Nice work, lgtm.

File: dot/network/notifications.go [hunk 3]
```diff
@@ -152,9 +152,12 @@ func createDecoder(info *notificationsProtocol, handshakeDecoder HandshakeDecode
...
 
 // createNotificationsMessageHandler returns a function that is called by the handler of *inbound* streams.
-func (s *Service) createNotificationsMessageHandler(info *notificationsProtocol,
+func (s *Service) createNotificationsMessageHandler(
+	info *notificationsProtocol,
 	messageHandler NotificationsMessageHandler,
-	batchHandler NotificationsMessageBatchHandler) messageHandler {
+	batchHandler NotificationsMessageBatchHandler,
+) messageHandler {
+
```
Comment (qdm12) [inline dot/network/notifications.go:160]: nit Let's get rid of the leading empty line perhaps

```suggestion
```

File: dot/network/notifications.go (hunk 1 above)
Comment (qdm12) [reply dot/network/notifications.go:162]: Do we want to close the stream in every return case :thinking: ? How about `CloseRead`?

Comment (EclesioMeloJunior) [reply dot/network/notifications.go:162]: hmm, got it!

**Review by EclesioMeloJunior — APPROVED** [review]

File: dot/network/notifications.go (hunk 1 above)
Comment (qdm12) [reply dot/network/notifications.go:162]: @EclesioMeloJunior how about `CloseRead`?

Comment (EclesioMeloJunior) [reply dot/network/notifications.go:162]: @qdm12 @kishansagathiya at line 354 from `dot/network/notifications.go` we open a stream, we could add a `defer` which will call the `CloseRead` when the function exits

Comment (kishansagathiya) [reply dot/network/notifications.go:161]: `messageHandler` returned by `createNotificationsMessageHandler` is used in `readStream`.

`readStream` always resets the stream in the end using defer. When you reset the stream, the stream gets closed from both ends.

https://github.com/ChainSafe/gossamer/blob/f091a756d24663000b4ca99ba39e4aafbde95e84/dot/network/inbound.go#L14

Please unresolve, if you feel this still needs to be addressed.

**Pull request merged**